### PR TITLE
Keep LayerNorm accumulator at FP32

### PR DESCRIPTION
**Pull request by lakhinderwalia**
Avoid the overflow in calculations of variance..

### CODE REVIEW

File: src/targets/gpu/kernels/include/migraphx/kernels/layernorm.hpp [hunk 1]
```diff
@@ -50,34 +50,28 @@ __device__ void generic_binary_layernorm(
...
+        auto mean  = x_sum / n;
 
-        r.inner([&](auto& y, auto x, auto... xs) {
-            auto m = x - mean_x;
+        auto x_diff = r.inner([&](auto x) { return x - mean; })(input);
 
-            // m * rsqrt(mean(m ^ 2) + epsilon)
-            y = compute(m * rsqrt(variance + eps_val), xs...);
-        })(output, input, inputs...);
+        auto sq_diff_sum = r.reduce(op::sum{}, 0, [](auto dx) { return dx * dx; })(x_diff);
```
Comment (pfultz2) [inline src/targets/gpu/kernels/include/migraphx/kernels/layernorm.hpp:65]: We dont want to do two reductions separately as that is much slower than doing the reductions together in parallel.

Comment (lakhinderwalia) [reply src/targets/gpu/kernels/include/migraphx/kernels/layernorm.hpp]: That was done for correctness, but it has been changed to just accumulation in higher precision, for speed reasons. Thanks.

File: src/targets/gpu/kernels/include/migraphx/kernels/layernorm.hpp [hunk 2]
```diff
@@ -50,33 +62,33 @@ __device__ void generic_binary_layernorm(
     using reduce_output = reduce::with_axis<Input1, Axis>;
 
     block::template run<reduce_output>([&](auto, auto r) {
-        auto input       = r.inner([&](auto x1, auto x2) { return op(x1, x2); })(input1, input2);
-        using value_type = typename Input1::type;
-        using vec_value_type       = vec_type<value_type>;
+        using value_type     = typename Input1::type;
+        using vec_value_type = typename acc_type<vec_type<value_type>>::type;
```
Comment (umangyadav) [inline src/targets/gpu/kernels/include/migraphx/kernels/layernorm.hpp:66]: nit :  `vec_value_type` is not meaningful name. This is accumulator type not the value tyepe of vectorized input.

**Comment by lakhinderwalia** [comment]
Test of test_layernorm_large with this change show very similar numbers: (on banff-cyxtera-s81-2)

With this change:
`# bin/test_verify test_layernorm_large                                                           [   RUN    ] test_layernorm_large                                                                                                    [ COMPLETE ] test_layernorm_large (12240.9ms)                                                                                        [==========] 1 tests ran
`                

Original code: 

`# bin/test_verify test_layernorm_large                                                           [   RUN    ] test_layernorm_large                                                                                                    [ COMPLETE ] test_layernorm_large (12226ms)                                                                                          [==========] 1 tests ran`

**Comment by umangyadav** [comment]
> Test of test_layernorm_large with this change show very similar numbers: (on banff-cyxtera-s81-2)
> 
> With this change: `# bin/test_verify test_layernorm_large [ RUN ] test_layernorm_large [ COMPLETE ] test_layernorm_large (12240.9ms) [==========] 1 tests ran `
> 
> Original code:
> 
> `# bin/test_verify test_layernorm_large [ RUN ] test_layernorm_large [ COMPLETE ] test_layernorm_large (12226ms) [==========] 1 tests ran`

Is this with Shapes ? 
        std::vector<size_t> dims = {1, 32, 8388608};
 ?

**Comment by lakhinderwalia** [comment]
> Is this with Shapes ? std::vector<size_t> dims = {1, 32, 8388608}; ?

This is with std::vector<size_t> dims = {1, 32, 262144};

> > Test of test_layernorm_large with this change show very similar numbers: (on banff-cyxtera-s81-2)
> > With this change: `# bin/test_verify test_layernorm_large [ RUN ] test_layernorm_large [ COMPLETE ] test_layernorm_large (12240.9ms) [==========] 1 tests ran `
> > Original code:
> > `# bin/test_verify test_layernorm_large [ RUN ] test_layernorm_large [ COMPLETE ] test_layernorm_large (12226ms) [==========] 1 tests ran`
> 
> Is this with Shapes ? std::vector<size_t> dims = {1, 32, 8388608}; ?

This is       std::vector<size_t> dims = {1, 32, 262144};
[test_layernorm_large](https://github.com/ROCm/AMDMIGraphX/blob/5450cfb8a6f6db59b296014bca3a41e06b57ba3b/test/verify/test_layernorm.cpp#L100)

**Comment by lakhinderwalia** [comment]
LayerNorm operator perf report comparison testing of a large FP16 tensor: 
`half_type, {1, 32, 134217728}`
 
Total time: 
`341.899 ms` (old code)  
`341.936 ms` (modified code, with fp32 accumulation).

**Review by CharlieL7 — APPROVED** [review]
LGTM

**Review by kahmed10 — APPROVED** [review]

**Pull request merged**